bazaarvoice/emodb
# Stop Writing Audits to Cassandra

## Github Issue #

None

## What Are We Doing Here?

Currently EmoDB writes all audits to Cassandra in the same mutation batch as it corresponding delta. The problem with this approach is that audits are not compacted in any way, so they grow infinitely has more writes occur in the system, regardless of whether they are updates to existing data.

This PR modifies our Emo to write audits in batches to S3, rather than to Cassandra. The benefit of this is that it is WAY cheaper to store data in S3, and also ensures that audit storage cannot interfere with delta storage, which is the data we are primarily concerned with. The downsides to this include decreased durability guarantees and slower audit retrieval times.

In doing this, it was required that the `getTimeline` functionality of Emo be modified, as audits can no longer be read through EmoDB. `getTimeline` requests now return `UnsupportedOperationException` when asked to include audit data. Lastly, users must use Athena queries to read audit data directly from S3.

## How to Test and Verify

1. Check out this PR
2. Update `config-local.yaml` with the bucket and path you would like to write to
3. Write some data to emo and observe that content is flushed to S3 every 2 minutes
4. Write some more data and then quickly shutdown emo using `curl -XPOST localhost:8081/tasks/graceful-shutdown` 
5. Observe that Emo gracefully shuts down and uploaded all remaining audits to s3 prior to termination

## Risk

### Level 

`High`

### Required Testing

`Regression` and `Manual`

### Risk Summary

Considering this fundamentally changes the way we store audit logs, this is pretty high risk. We need to verify not only the durability of the audits, but also that unforseen issues with s3 does not block the writes of deltas.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
